### ISSUE TITLE
Set sysctl net/ipv4/ip_unprivileged_port_start ?

### ISSUE THREAD

**Issue by thockin** [kubernetes/kubernetes#102612]
This would remove one more reason why pods run as root when they otherwise don't need to.  Clearly this is a CRI thing, but perhaps we should require it?

**Comment by pacoxu** [comment]
Add the moby PR link and also containerd issue: https://github.com/containerd/containerd/issues/4936
- Done: Add default sysctls to allow ping sockets and privileged ports with no capabilities https://github.com/moby/moby/pull/41030 
- WIP moby:  Remove CAP_NET_RAW from default capabilities https://github.com/moby/moby/issues/41886

**Comment by ehashman** [comment]
Looks like this is effectively a dupe of #56374

Closing this in favour of that issue

/close

**Comment by ehashman** [comment]
Hrm, attempting to summarize that one makes me wonder if I spoke too soon...

I think there's a couple things going on here: we need to figure out if we want to add some new sysctls to the allow-list, and we also should figure out a plan for how to deal with the issue described in the linked bug. I'm not sure how much overlap there is yet, so I will reopen this.

/reopen

**Comment by BenTheElder** [comment]
relevant comment https://github.com/kubernetes/kubernetes/issues/56374#issuecomment-868892588

**Comment by mrunalp** [comment]
Yes, we allow setting this by default in CRI-O and docker does it, too IIRC. It makes sense to enable this by default :+1:

**Comment by olljanat** [comment]
> Yes, we allow setting this by default in CRI-O and docker does it, too IIRC. It makes sense to enable this by default 👍

@mrunalp where CRI-O allows this one? I see that crio.conf supports `default_sysctls` but it have only `ping_group_range` at least on documentation. Instead of there looks to be `NET_BIND_SERVICE` capability by default but it is little bit different thing.

**Comment by olljanat** [comment]
> FYI. I created PR to add this to containerd side same way like it is done on Docker [containerd/containerd#6170](https://github.com/containerd/containerd/pull/6170)

Feedback looks to be that those Docker changes caused some unexpected issues which why there most likely will be two new options `enable_unprivileged_icmp` and `enable_unprivileged_ports` on containerd which defaults to `false`.

**Comment by BenTheElder** [comment]
> Kubernetes marked it as a safe sysctl since v1.22 https://github.com/kubernetes/kubernetes/pull/103326.

Yes, but doing it by default would probably help more people avoid over-permission-ing their pods

**Comment by pacoxu** [comment]
- https://github.com/kubernetes/kubernetes/issues/103298 was done in v1.22
- containerd plan to do it in v2.0 https://github.com/containerd/containerd/issues/6924

**Comment by pacoxu** [comment]
https://github.com/containerd/containerd/pull/9348 is merged which will be included in containerd v2.0 which is not released yet.


https://github.com/kubernetes/kubernetes/pull/105309#issuecomment-1635303233 this needs kernel 4.11+

Or coredns pod will fail with below error:

Warning FailedCreatePodSandBox 2s (x13 over 43s) kubelet Failed to create pod sandbox: rpc error: code = Unknown desc = failed to create containerd task: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error during container init: open /proc/sys/net/ipv4/ip_unprivileged_port_start: no such file or directory: unknown

For more context, see https://github.com/kubernetes/kubernetes/pull/105309#issuecomment-1635303233.


The coredns change may wait for minimal kernel change to 4.11+.

**Comment by thockin** [comment]
Once those land in containerd 2.0, does kubernetes need to do anything?  Or do we get it for free?

Will CRI-O match?

**Comment by pacoxu** [comment]
**Will CRI-O match?**

> Yes, we allow setting this by default in CRI-O and docker does it, too IIRC. It makes sense to enable this by default 👍

CRI-O matches this by default.

> Once those land in containerd 2.0, does kubernetes need to do anything? Or do we get it for free?

As a safe sysctl, it's free for kubernetes users. However, the kernel version is a problem here.

This sysctl needs kernel 4.11+. And the check can be done in kubeadm side.

For new clusters:
- In system-validator, https://github.com/kubernetes/system-validators/pull/37 was merged recently.
- [Action Item] We need to bump it in k/k  later. 

[Action Item] For upgrades, I think there is no kernel version check there. I need to verify to see if we need to add a warning for upgrading.

**Comment by thockin** [comment]
I guess what I meant is: will we set this by default?  Is there any reason NOT to set it?

**Comment by pacoxu** [comment]
>> Once those land in containerd 2.0, does kubernetes need to do anything? Or do we get it for free?
> 
>> Will CRI-O match?

> I guess what I meant is: will we set this by default? Is there any reason NOT to set it?

IIUC, in https://github.com/cri-o/cri-o/issues/6706, @haircommander mentions that this can be done by adding it to the default_sysctls. This still needs to be configured. 

Which means: 
- Containerd 2.0 sets this sysctl by default
- CRI-O does not.

Still, the kernel version requirement is a prerequisite.

**Comment by tianon** [comment]
See https://github.com/moby/moby/pull/41030/files#diff-9f91bff23e0bd70d6429b63d9db2d8180d2e89cdb64db4fb3e10a96f74d36271R782 for the cases Docker excludes by default -- basically host network and other pre-existing network namespaces; any _new_ network namespaces gets them by default.

**Comment by thockin** [comment]
Thanks!  That seems like a great end-state

**Comment by BenTheElder** [comment]
Note: https://github.com/containerd/containerd/pull/11502

Containerd 2.0 did ship with enabling unprivileged ports and unprivileged ping for CRI / pods by default.

(After thread pulling ... https://github.com/kubernetes-sigs/kind/issues/3876 ... recalled this issue)

It sounds like cri-dockerd may implicitly as well? But not cri-o.

**Comment by SergeyKanzhelev** [comment]
> Note: [containerd/containerd#11502](https://github.com/containerd/containerd/pull/11502)
> 
> Containerd 2.0 did ship with enabling unprivileged ports and unprivileged ping for CRI / pods by default.
> 
> (After thread pulling ... [kubernetes-sigs/kind#3876](https://github.com/kubernetes-sigs/kind/issues/3876) ... recalled this issue)
> 
> It sounds like cri-dockerd may implicitly as well? But not cri-o.

@haircommander something you want to track for cri-o?
